File: 2021/ruby/day2.rb
#! /usr/bin/env ruby

input = <<INPUT
forward 5
down 5
forward 8
up 3
down 8
forward 2
INPUT

data = input.split("\n")
# data = File.read("../inputs/day2t").split("\n")
# data = File.read("../inputs/day2").split("\n")

class Parser
  def initialize(row)
    comm, val = row.split
    @comm = comm.to_sym
    @val = val.to_i
  end

  def get
    return @comm, @val
  end

  def deconstruct_keys(keys)
    Hash[@comm, @val]
  end
end

class Day2
  def initialize(data)
    @data = data
  end

  def run1
    res = @data.inject({ h: 0, v: 0 }) { |memo, row| task1(memo, row) }
    puts("res1: #{res[:h] * res[:v]}")
  end

  # well, ruby pattern matching suck...
  def task1(memo, row)
    case Parser.new(row)
    in forward: Integer => dig
      memo[:h] += dig
    in up: Integer => dig
      memo[:v] -= dig
    in down: Integer => dig
      memo[:v] += dig
    else
      :no_match
    end
    memo
  end
end

def task2(memo, row)
  com, dig = Parser.new(row).get

  case com
  when :forward
    memo[:h] += dig
    memo[:d] += memo[:aim] * dig
  when :up
    memo[:aim] -= dig
  when :down
    memo[:aim] += dig
  end

  memo
end

Day2.new(data).run1

res2 = data.inject({ h: 0, d: 0, aim: 0 }) { |memo, row| task2(memo, row) }
puts("res2: #{res2[:h] * res2[:d]}")

File: 2021/elixir/day2.exs
# day 2 is much more fun with functional programming

defmodule Day2 do
  def parse_all(fname) do
    fname
    |> File.stream!()
    |> Stream.map(&String.trim/1)
    |> Stream.map(&parse/1)
  end

  def task1({:u, val}, acc), do: Map.update!(acc, :v, &(&1 - val))
  def task1({:d, val}, acc), do: Map.update!(acc, :v, &(&1 + val))
  def task1({:f, val}, acc), do: Map.update!(acc, :h, &(&1 + val))

  def task2({:u, val}, acc), do: Map.update!(acc, :aim, &(&1 - val))
  def task2({:d, val}, acc), do: Map.update!(acc, :aim, &(&1 + val))

  def task2({:f, val}, acc) do
    acc = Map.update!(acc, :h, &(&1 + val))
    Map.update!(acc, :v, &(&1 + val * acc.aim))
  end

  defp parse(<<"forward ", val::binary>>), do: {:f, to_i(val)}
  defp parse(<<"up ", val::binary>>), do: {:u, to_i(val)}
  defp parse(<<"down ", val::binary>>), do: {:d, to_i(val)}

  defp to_i(string), do: String.to_integer(string)
end

input = "../inputs/day2t"
# input = "../inputs/day2"

# task 1
input
|> Day2.parse_all()
|> Enum.reduce(%{v: 0, h: 0}, &Day2.task1/2)
|> Map.values()
|> Enum.product()
|> IO.inspect(label: "res1")

# task 2
input
|> Day2.parse_all()
|> Enum.reduce(%{v: 0, h: 0, aim: 0}, &Day2.task2/2)
|> Map.take([:h, :v])
|> Map.values()
|> Enum.product()
|> IO.inspect(label: "res2")
